controllers/customer_controller.go

```diff
@@ -0,0 +1,203 @@
+
+		customer := r.JiraServiceDeskClient.GetCustomerFromCustomerCRForCreateCustomer(instance)
+		customerId, err := r.JiraServiceDeskClient.CreateCustomer(customer)
+		if err != nil {
+			return reconcilerUtil.ManageError(r.Client, instance, err, false)
+		}
+
+		instance.Status.AccountId = customerId
+		log.Info("Successfully created Jira Service Desk Customer: " + instance.Spec.DisplayName)
+	}
+
+	for _, projectKey := range instance.Spec.ProjectKeys {
+		if !contains(instance.Status.ProjectKeys, projectKey) {
+			err := r.JiraServiceDeskClient.AddCustomerToProject(instance.Status.AccountId, projectKey)
+			if err != nil {
+				return reconcilerUtil.ManageError(r.Client, instance, err, false)
+			}
+			instance.Status.ProjectKeys = append(instance.Status.ProjectKeys, projectKey)
+			log.Info("Successfully added Jira Service Desk Customer into project: " + projectKey)
+		}
+	}
+
+	for index, projectKey := range instance.Status.ProjectKeys {
+		if !contains(instance.Spec.ProjectKeys, projectKey) {
+			err := r.JiraServiceDeskClient.RemoveCustomerFromProject(instance.Status.AccountId, projectKey)
+			if err != nil {
+				return reconcilerUtil.ManageError(r.Client, instance, err, false)
+			}
+			instance.Status.ProjectKeys[index] = ""
+			log.Info("Successfully removed Jira Service Desk Customer from project: " + projectKey)
+		}
+	}
+
+	instance.Status.ProjectKeys = deleteEmpty(instance.Status.ProjectKeys)
+
+	return reconcilerUtil.ManageSuccess(r.Client, instance)
+}
+
+func (r *CustomerReconciler) handleDelete(req ctrl.Request, instance *jiraservicedeskv1alpha1.Customer) (ctrl.Result, error) {
+	log := r.Log.WithValues("customer", req.NamespacedName)
+
+	if instance == nil {
+		// Instance not found, nothing to do
+		return reconcilerUtil.DoNotRequeue()
+	}
+
+	log.Info("Removing Jira Service Desk Customer: " + instance.Spec.DisplayName)
+
+	// Remove customer from JSD project
+	for _, projectKey := range instance.Status.ProjectKeys {
+		err := r.JiraServiceDeskClient.RemoveCustomerFromProject(instance.Status.AccountId, projectKey)
+		if err != nil {
+			return reconcilerUtil.ManageError(r.Client, instance, err, false)
+		}
+		log.Info("Successfully removed Jira Service Desk Customer from project: " + projectKey)
+	}
+
+	// Delete Finalizer
+	finalizerUtil.DeleteFinalizer(instance, CustomerFinalizer)
+
+	log.Info("Finalizer removed for customer: " + instance.Spec.DisplayName)
+
+	// Update instance
+	err := r.Client.Update(context.TODO(), instance)
+	if err != nil {
+		return reconcilerUtil.ManageError(r.Client, instance, err, false)
+	}
+
+	return reconcilerUtil.DoNotRequeue()
+}
+
+func contains(slice []string, search string) bool {
+	for _, value := range slice {
+		if value == search {
+			return true
+		}
+	}
+	return false
+}
+
+func deleteEmpty(slice []string) []string {
+	var output []string
+	for _, str := range slice {
+		if str != "" {
+			output = append(output, str)
+		}
+	}
+	return output
```

Comment:
I am not sure why these methods are required

Reply:
- Needs contain to check if a project in Customer.Status.AssociatedProjects is present in Customer.Spec.Projects and vice versa. The code was used at multiple places so converted it into a function
- While removing the associated projects, had to cleanup the project.Status.AssociatedProjects slice. So emptied the slice first with an empty string and later removed all the reserved space via the deleteEmpty function().

Reply:
For the whole functionality of comparing two arrays please use a more generic approach. Add a method that can compare two arrays and then use that. 
Something like https://play.golang.org/p/KHTmJcR4rg